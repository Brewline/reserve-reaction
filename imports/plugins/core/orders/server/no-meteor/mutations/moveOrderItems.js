import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import { Order as OrderSchema } from "/imports/collections/schemas";
import updateGroupStatusFromItemStatus from "../util/updateGroupStatusFromItemStatus";
import updateGroupTotals from "../util/updateGroupTotals";

// These should eventually be configurable in settings
const itemStatusesThatOrdererCanMove = ["new"];
const orderStatusesThatOrdererCanMove = ["new"];

const inputSchema = new SimpleSchema({
  "fromFulfillmentGroupId": String,
  "itemIds": {
    type: Array,
    minCount: 1
  },
  "itemIds.$": String,
  "orderId": String,
  "toFulfillmentGroupId": String
});

/**
 * @method moveOrderItems
 * @summary Use this mutation to move one or more items between existing order
 *   fulfillment groups.
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Necessary input. See SimpleSchema
 * @return {Promise<Object>} Object with `order` property containing the updated order
 */
export default async function moveOrderItems(context, input) {
  inputSchema.validate(input);

  const {
    fromFulfillmentGroupId,
    itemIds,
    orderId,
    toFulfillmentGroupId
  } = input;

  const { accountId, appEvents, collections, isInternalCall, userHasPermission, userId } = context;
  const { Orders } = collections;

  // First verify that this order actually exists
  const order = await Orders.findOne({ _id: orderId });
  if (!order) throw new ReactionError("not-found", "Order not found");

  // Allow move if the account that placed the order is attempting to move
  // or if the account has "orders" permission. When called internally by another
  // plugin, context.isInternalCall can be set to `true` to disable this check.
  if (
    !isInternalCall &&
    (!accountId || accountId !== order.accountId) &&
    !userHasPermission(["orders"], order.shopId)
  ) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  // Is the account calling this mutation also the account that placed the order?
  // We need this check in a couple places below, so we'll get it here.
  const accountIsOrderer = (order.accountId && accountId === order.accountId);

  // The orderer may only move items while the order status is still "new"
  if (accountIsOrderer && !orderStatusesThatOrdererCanMove.includes(order.workflow.status)) {
    throw new ReactionError("invalid", `Order status (${order.workflow.status}) is not one of: ${orderStatusesThatOrdererCanMove.join(", ")}`);
  }

  // Find the two fulfillment groups we're modifying
  const fromGroup = order.shipping.find((group) => group._id === fromFulfillmentGroupId);
  if (!fromGroup) throw new ReactionError("not-found", "Order fulfillment group (from) not found");

  const toGroup = order.shipping.find((group) => group._id === toFulfillmentGroupId);
  if (!toGroup) throw new ReactionError("not-found", "Order fulfillment group (to) not found");

  // Pull out the item's we're moving
  const foundItemIds = [];
  const movedItems = fromGroup.items.reduce((list, item) => {
    if (itemIds.includes(item._id)) {
      // The orderer may only move while the order item status is still "new"
      if (accountIsOrderer && !itemStatusesThatOrdererCanMove.includes(item.workflow.status)) {
        throw new ReactionError("invalid", `Item status (${item.workflow.status}) is not one of: ${itemStatusesThatOrdererCanMove.join(", ")}`);
      }

      list.push(item);
      foundItemIds.push(item._id);
    }
    return list;
  }, []);

  if (!itemIds.every((id) => foundItemIds.includes(id))) {
    throw new ReactionError("not-found", "Some order items not found");
  }

  const { billingAddress, cartId, currencyCode } = order;

  // Find and move the items
  const orderSurcharges = [];
  const updatedGroups = await Promise.all(order.shipping.map(async (group) => {
    if (group._id !== fromFulfillmentGroupId && group._id !== toFulfillmentGroupId) return group;

    let updatedItems;
    if (group._id === fromFulfillmentGroupId) {
      // Remove the moved items
      updatedItems = group.items.filter((item) => !itemIds.includes(item._id));
    } else {
      // Add the moved items
      updatedItems = [...group.items, ...movedItems];
    }

    if (updatedItems.length === 0) {
      throw new ReactionError("invalid-param", "move would result in group having no items");
    }

    // Create an updated group
    const updatedGroup = {
      ...group,
      // There is a convenience itemIds prop, so update that, too
      itemIds: updatedItems.map((item) => item._id),
      items: updatedItems,
      totalItemQuantity: updatedItems.reduce((sum, item) => sum + item.quantity, 0)
    };

    // Update group shipping, tax, totals, etc.
    const { groupSurcharges } = await updateGroupTotals(context, {
      billingAddress,
      cartId,
      currencyCode,
      discountTotal: updatedGroup.invoice.discounts,
      group: updatedGroup,
      orderId,
      selectedFulfillmentMethodId: updatedGroup.shipmentMethod._id
    });

    // Push all group surcharges to overall order surcharge array.
    // Currently, we do not save surcharges per group
    orderSurcharges.push(...groupSurcharges);

    // Ensure proper group status
    updateGroupStatusFromItemStatus(updatedGroup);

    return updatedGroup;
  }));

  // We're now ready to actually update the database and emit events
  const modifier = {
    $set: {
      shipping: updatedGroups,
      surcharges: orderSurcharges,
      totalItemQuantity: updatedGroups.reduce((sum, group) => sum + group.totalItemQuantity, 0),
      updatedAt: new Date()
    }
  };

  OrderSchema.validate(modifier, { modifier: true });

  const { modifiedCount, value: updatedOrder } = await Orders.findOneAndUpdate(
    { _id: orderId },
    modifier,
    { returnOriginal: false }
  );
  if (modifiedCount === 0 || !updatedOrder) throw new ReactionError("server-error", "Unable to update order");

  await appEvents.emit("afterOrderUpdate", {
    order: updatedOrder,
    updatedBy: userId
  });

  return { order: updatedOrder };
}
