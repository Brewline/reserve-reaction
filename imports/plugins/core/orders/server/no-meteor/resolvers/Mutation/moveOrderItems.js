import {
  decodeOrderFulfillmentGroupOpaqueId,
  decodeOrderItemOpaqueId,
  decodeOrderOpaqueId
} from "@reactioncommerce/reaction-graphql-xforms/order";

/**
 * @name Mutation/moveOrderItems
 * @method
 * @memberof Payments/GraphQL
 * @summary resolver for the moveOrderItems GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.fromFulfillmentGroupId - The ID of the order fulfillment group from which all the items
 *   are to be moved.
 * @param {String[]} args.input.itemIds - The list of item IDs to move. The full quantity must be moved.
 * @param {String} args.input.orderId - The order ID
 * @param {String} args.input.toFulfillmentGroupId - The ID of the order fulfillment group to which all the items
 *   are to be moved.
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} MoveOrderItemsPayload
 */
export default async function moveOrderItems(parentResult, { input }, context) {
  const {
    clientMutationId = null,
    fromFulfillmentGroupId,
    itemIds,
    orderId,
    toFulfillmentGroupId
  } = input;

  const { order } = await context.mutations.moveOrderItems(context, {
    fromFulfillmentGroupId: decodeOrderFulfillmentGroupOpaqueId(fromFulfillmentGroupId),
    itemIds: itemIds.map(decodeOrderItemOpaqueId),
    orderId: decodeOrderOpaqueId(orderId),
    toFulfillmentGroupId: decodeOrderFulfillmentGroupOpaqueId(toFulfillmentGroupId)
  });

  return {
    clientMutationId,
    order
  };
}
