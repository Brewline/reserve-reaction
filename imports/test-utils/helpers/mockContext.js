const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
