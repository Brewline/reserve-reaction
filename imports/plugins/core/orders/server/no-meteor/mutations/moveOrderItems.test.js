import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import {
  restore as restore$updateGroupTotals,
  rewire as rewire$updateGroupTotals
} from "../util/updateGroupTotals";
import moveOrderItems from "./moveOrderItems";

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(() => {
  restore$updateGroupTotals();
});

test("throws if orderId isn't supplied", async () => {
  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if itemIds isn't supplied", async () => {
  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if itemIds is empty", async () => {
  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: [],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if fromFulfillmentGroupId isn't supplied", async () => {
  await expect(moveOrderItems(mockContext, {
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if toFulfillmentGroupId isn't supplied", async () => {
  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(null));

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the fromFulfillmentGroup doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        _id: "group200",
        items: [
          {
            _id: "xyz",
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the toFulfillmentGroup doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        _id: "group1",
        items: [
          {
            _id: "xyz",
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if an order item doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        _id: "group1",
        items: [
          {
            _id: "xyz",
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      },
      {
        _id: "group2",
        items: [
          {
            _id: "xyz",
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if permission check fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: []
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(false);

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders"], "SHOP_ID");
});

test("throws if user who placed order tries to move item at invalid current item status", async () => {
  const accountId = "ACCOUNT_ID";

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    accountId,
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group1",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item1",
            quantity: 1,
            workflow: {
              status: "processing",
              workflow: ["new", "processing"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group2",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item2",
            quantity: 1,
            workflow: {
              status: "processing",
              workflow: ["new", "processing"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.accountId = accountId;

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();

  mockContext.accountId = null;
});

test("throws if user who placed order tries to move item at invalid current order status", async () => {
  const accountId = "ACCOUNT_ID";

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    accountId,
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group1",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item1",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group2",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item2",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "processing",
      workflow: ["new", "processing"]
    }
  }));

  mockContext.accountId = accountId;

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();

  mockContext.accountId = null;
});

test("throws if the from group would have no items remaining", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group1",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item1",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group2",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item2",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 0
  }));

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the database update fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group1",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item1",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }),
          Factory.OrderItem.makeOne({
            _id: "item10",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group2",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item2",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 0
  }));

  await expect(moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("skips permission check if context.isInternalCall", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group1",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item1",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }),
          Factory.OrderItem.makeOne({
            _id: "item10",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "group2",
        items: [
          Factory.OrderItem.makeOne({
            _id: "item2",
            quantity: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  mockContext.isInternalCall = true;

  const mockUpdateGroupTotals = jest.fn().mockName("updateGroupTotals").mockReturnValue(Promise.resolve({ groupSurcharges: [] }));
  rewire$updateGroupTotals(mockUpdateGroupTotals);

  await moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  });

  delete mockContext.isInternalCall;

  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
});

test("moves items", async () => {
  const group1Items = Factory.OrderItem.makeMany(3, {
    _id: (index) => `item_1_${index}`,
    quantity: 1,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const group2Items = Factory.OrderItem.makeMany(2, {
    _id: (index) => `item_2_${index}`,
    quantity: 1,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const group1 = Factory.OrderFulfillmentGroup.makeOne({
    _id: "group1",
    items: group1Items,
    totalItemQuantity: group1Items.reduce((sum, item) => sum + item.quantity, 0)
  });

  const group2 = Factory.OrderFulfillmentGroup.makeOne({
    _id: "group2",
    items: group2Items,
    totalItemQuantity: group2Items.reduce((sum, item) => sum + item.quantity, 0)
  });

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "order1",
    shipping: [group1, group2],
    shopId: "SHOP_ID",
    totalItemQuantity: [group1, group2].reduce((sum, group) => sum + group.totalItemQuantity, 0),
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  const mockUpdateGroupTotals = jest.fn().mockName("updateGroupTotals").mockReturnValue(Promise.resolve({ groupSurcharges: [] }));
  rewire$updateGroupTotals(mockUpdateGroupTotals);

  await moveOrderItems(mockContext, {
    fromFulfillmentGroupId: "group1",
    itemIds: ["item_1_0", "item_1_1"],
    orderId: "order1",
    toFulfillmentGroupId: "group2"
  });

  const expectedGroup2Items = [...group2.items, group1.items[0], group1.items[1]];

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "order1" },
    {
      $set: {
        shipping: [
          {
            ...group1,
            items: [group1.items[2]],
            itemIds: [group1.items[2]._id],
            totalItemQuantity: 1
          },
          {
            ...group2,
            items: expectedGroup2Items,
            itemIds: expectedGroup2Items.map((item) => item._id),
            totalItemQuantity: 4
          }
        ],
        surcharges: [],
        totalItemQuantity: 5,
        updatedAt: jasmine.any(Date)
      }
    },
    { returnOriginal: false }
  );
});
