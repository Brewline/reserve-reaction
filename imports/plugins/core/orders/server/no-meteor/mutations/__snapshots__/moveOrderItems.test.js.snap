// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if an order item doesn't exist 1`] = `"Some order items not found"`;

exports[`throws if fromFulfillmentGroupId isn't supplied 1`] = `"From fulfillment group ID is required"`;

exports[`throws if itemIds is empty 1`] = `"You must specify at least 1 values"`;

exports[`throws if itemIds isn't supplied 1`] = `"Item ids is required"`;

exports[`throws if orderId isn't supplied 1`] = `"Order ID is required"`;

exports[`throws if permission check fails 1`] = `"Access Denied"`;

exports[`throws if the database update fails 1`] = `"queries.getFulfillmentMethodsWithQuotes is not a function"`;

exports[`throws if the from group would have no items remaining 1`] = `"move would result in group having no items"`;

exports[`throws if the fromFulfillmentGroup doesn't exist 1`] = `"Order fulfillment group (from) not found"`;

exports[`throws if the order doesn't exist 1`] = `"Order not found"`;

exports[`throws if the toFulfillmentGroup doesn't exist 1`] = `"Order fulfillment group (to) not found"`;

exports[`throws if toFulfillmentGroupId isn't supplied 1`] = `"To fulfillment group ID is required"`;

exports[`throws if user who placed order tries to move item at invalid current item status 1`] = `"Item status (processing) is not one of: new"`;

exports[`throws if user who placed order tries to move item at invalid current order status 1`] = `"Order status (processing) is not one of: new"`;
